pkg/utils/cliutils/printer.go

```diff
@@ -0,0 +1,124 @@
+package cliutils
+
+import (
+	"fmt"
+	"github.com/solo-io/solo-kit/pkg/utils/protoutils"
+	"io"
+	"reflect"
+	"strings"
+	"text/template"
+
+	"github.com/ghodss/yaml"
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+)
+
+// Printer represents a function that prints a value to io.Writer, usually using
+// a table
+type Printer func(interface{}, io.Writer) error
+
+// Print - prints the given proto.Message to io.Writer using the specified output format
+func Print(output, template string, m proto.Message, tblPrn Printer, w io.Writer) error {
+	switch strings.ToLower(output) {
+	case "yaml":
+		return PrintYAML(m, w)
+	case "json":
+		return PrintJSON(m, w)
+	case "template":
+		return PrintTemplate(m, template, w)
+	default:
+		return tblPrn(m, w)
+	}
+}
+
+// PrintList - prints the given list of values to io.Writer using the specified output format
+func PrintList(output, template string, list interface{}, tblPrn Printer, w io.Writer) error {
+	switch strings.ToLower(output) {
+	case "yaml":
+		return PrintYAMLList(list, w)
+	case "json":
+		return PrintJSONList(list, w)
+	case "template":
+		return PrintTemplate(list, template, w)
+	default:
+		return tblPrn(list, w)
+	}
+}
+
+// PrintJSON - prints the given proto.Message to io.Writer in JSON
+func PrintJSON(m proto.Message, w io.Writer) error {
+	b, err := protoutils.MarshalJSON(m)
+	if err != nil {
+		return errors.Wrap(err, "unable to convert to JSON")
+	}
+	_, err = fmt.Fprintln(w, string(b))
+	return err
+}
+
+// PrintYAML - prints the given proto.Message to io.Writer in YAML
+func PrintYAML(m proto.Message, w io.Writer) error {
+	jsn, err := protoutils.MarshalJSON(m)
+	if err != nil {
+		return errors.Wrap(err, "unable to marshal")
+	}
+	b, err := yaml.JSONToYAML(jsn)
+	if err != nil {
+		return errors.Wrap(err, "unable to convert to YAML")
+	}
+	_, err = fmt.Fprintln(w, string(b))
+	return err
+}
+
+// PrintJSONList - prints the given list to io.Writer in JSON
+func PrintJSONList(data interface{}, w io.Writer) error {
+	list := reflect.ValueOf(data)
+	_, err := fmt.Fprintln(w, "[")
+	if err != nil {
+		return errors.Wrap(err, "unable to print JSON list")
+	}
+	for i := 0; i < list.Len(); i++ {
+		v, ok := list.Index(i).Interface().(proto.Message)
+		if !ok {
+			return errors.New("unable to convert to proto message")
+		}
+		if i != 0 {
+			_, err = fmt.Fprintln(w, ",")
+			if err != nil {
+				return errors.Wrap(err, "unable to print JSON list")
+			}
+		}
+		err = PrintJSON(v, w)
+		if err != nil {
+			return err
+		}
+	}
+	_, err = fmt.Fprintln(w, "]")
+	return err
+}
+
+// PrintYAMLList - prints the given list to io.Writer in YAML
+func PrintYAMLList(data interface{}, w io.Writer) error {
```

Comment:
as above, we should not need to give yaml lists special treatment - it's still yaml

Reply:
True, but I think they have slightly disparate purposes. One is meant for an arbitary list while the other is meant specifically for a proto message, the list utils could probably go in a different file if anything, but I think they could have some use.

If we massage the data at all then we would most likely want to call the list functions afterwards as opposed to the proto functions.